File: 백준/Gold/4386. 별자리 만들기/별자리 만들기.kt
import java.util.*
import kotlin.math.pow
import kotlin.math.sqrt

fun main() = with(System.`in`.bufferedReader()) {
    val n = readLine().toInt()
    val coords = Array(n) { readLine().split(' ').map(String::toDouble) }
    close()

    val visit = BooleanArray(n)
    val heap = PriorityQueue<Star> { a, b -> a.dist.compareTo(b.dist) }
    heap.add(Star(0.0, 0))
    var net = n
    var result = 0.0

    while (net > 0) {
        val (dist, u) = heap.poll()
        if (visit[u]) continue
        visit[u] = true
        result += sqrt(dist)
        net--

        for (v in 1 until n) {
            if (visit[v]) continue
            heap.add(Star(dist(coords[u], coords[v]), v))
        }
    }
    print(result)
}

fun dist(u: List<Double>, v: List<Double>) = (v[0] - u[0]).pow(2) + (v[1] - u[1]).pow(2)

data class Star(val dist: Double, val idx: Int)

File: 백준/Gold/4386. 별자리 만들기/별자리 만들기.py
import sys
from math import sqrt

readline = sys.stdin.readline

n = int(readline())
stars = [tuple(map(float, readline().split(' '))) for _ in range(n)]

edges = []
for i in range(n - 1):
    for j in range(i + 1, n):
        dist = (stars[i][0] - stars[j][0]) ** 2 + (stars[i][1] - stars[j][1]) ** 2
        edges.append((dist, i, j))
edges.sort()

star_sets = [{i} for i in range(n)]
res = 0.0
for dist, u, v in edges:
    if star_sets[u] is star_sets[v]:
        continue
    res += sqrt(dist)
    union = star_sets[u].union(star_sets[v])
    if len(union) == n:
        break
    for i in union:
        star_sets[i] = union
print(res)